BlockChainRunner.java

```diff
@@ -0,0 +1,354 @@
+            );
+            if (sumAlreadySpend > sumToSpend) {
+                outputs.add(
+                    TransactionOutput.builder().value(sumAlreadySpend - sumToSpend).destPublicKey(userKey).build()
+                );
+            }
+
+            pendingTransactions.add(new Transaction(inputs, outputs));
+        });
+    }
+
+    private void addPendingTransactionsToPool()
+        throws NoSuchAlgorithmException, SignatureException, InvalidKeyException
+    {
+        Signature sign = Signature.getInstance("SHA256withDSA");
+
+        for (Transaction pendingTransaction : pendingTransactions) {
+            boolean isValid = true;
+            List<Pair<PublicKey, UnspentOutput>> mappingsToRemove = new ArrayList<>();
+
+            // Validate transaction here ↓
+            int transactionInputValue = 0;
+
+            if (!pendingTransaction.getTxId().equals(pendingTransaction.computeTransactionId())) {
+                log(String.format(
+                    "INFO: Invalid transaction %s: invalid transaction hash", pendingTransaction.getTxId()
+                ));
+                continue;
+            }
+
+            for (TransactionInput pendingInput : pendingTransaction.getInputs()) {
+                String sourceTxId = pendingInput.getTxId();
+                int sourceOutputIndex = pendingInput.getIndex();
+
+                Transaction sourceTransaction = getTransactionFromBlockChainById(sourceTxId);
+                TransactionOutput sourceOutput = sourceTransaction.getOutputs().get(sourceOutputIndex);
+
+                PublicKey sourcePk = sourceOutput.getDestPublicKey();
+                List<UnspentOutput> userUnspentOutputs = new ArrayList<>(unspentOutputMap.get(sourcePk));
+                Optional<UnspentOutput> unspentOutput = userUnspentOutputs.stream()
+                    .filter(output -> output.getTxId().equals(sourceTxId) && output.getOutputIndex() == sourceOutputIndex)
+                    .findFirst();
+
+                if (unspentOutput.isEmpty()) {
+                    log(String.format(
+                        "INFO: Invalid transaction %s: no unspent output present", pendingTransaction.getTxId()
+                    ));
+                    isValid = false;
+                    break;
+                }
+
+                sign.initVerify(sourceOutput.getDestPublicKey());
+                sign.update(sourceTransaction.getTxId().getBytes());
+                if (!sign.verify(pendingInput.getSignedTxId())) {
+                    log(String.format(
+                        "INFO: Invalid transaction %s: Invalid digital signature", pendingTransaction.getTxId()
+                    ));
+                    isValid = false;
+                    break;
+                }
+
+                mappingsToRemove.add(Pair.of(sourcePk, unspentOutput.get()));
+                transactionInputValue += sourceOutput.getValue();
+            }
+            if (!isValid) continue;
```

Comment:
In the lines 288 and 298 you are changing _isValid_ value to false and breaking, why would you cross-verify the same variable once more to continue throughout the code if you cannot pass it if value of the variable changes to _False_?

Would be advisable to scrap the if clause itself.

Reply:
The variable comes from the outer scope and it's value could have been changed in the previous iteration